### PR TITLE
Add middleware to warn for unapplied migrations

### LINKED ISSUE #1696: Middleware to warn for unapplied migrations
Not applying migrations can break upgrades and we've already seen installations out in the wild which have upgraded their docker containers without applying the migrations. This can result in serious data loss going forward if users do not upgrade their installations properly.

Django already contains the base functionality that we need: 

`./manage.py showmigrations` will show empty boxes `[ ]` in case a migration has not been applied.

It is possible to use `call_command` (see tcms/core/tests/test_set_domain.py) and then parse the output but preferably we can hunt down a specific function in Django that lists all of the migrations and can reuse it instead of relying on text parsing.

Then all of that should be bundled in a middleware, similar to `CheckSettingsMiddleware` (or even inside of it) and if there are unapplied migrations add warning to the request. It will be displayed in the browser.

**Definition of Done**:
- Middleware implemented, preferably without using `call_command`. If this isn't possible directly I need explanation & justification in the PR discussion how Django counts the migrations and which are applied/which aren't and why it won't be possible to use Django's internals for what we need.
- Tests added for the new middleware
- CI pass
- All code review passed & PR merged

Hints:
- Devel getting started docs: https://kiwitcms.readthedocs.io/en/latest/contribution.html#testing
- Django testing documentation: https://docs.djangoproject.com/en/3.0/topics/testing/
- Python unittest documentation: https://docs.python.org/3.7/library/unittest.html
- Check `django.core.management.base.check_migrations()`

*This issue is part of Kiwi TCMS open source bounty program. For more information see the link(s) in [bounty-program milestone](https://github.com/kiwitcms/Kiwi/milestone/4)*

### PR DESCRIPTION
**Pull request by brymut**
Resolves #1696. Add middlware to warn for unapplied migrations

### CODE REVIEW

**Comment by brymut** [comment]
Used the `django.db.migrations.executor.MigrationExecutor` as its used in [django.core.management.base.check_migrations()](https://github.com/django/django/blob/c60524c658f197f645b638f9bcc553103bfe2630/django/core/management/base.py#L475) will that work?

File: tcms/core/middleware.py [hunk 1]
```diff
@@ -32,3 +34,20 @@ def process_request(self, request):
...
+    def process_request(self, request):
+        executor = MigrationExecutor(connections[DEFAULT_DB_ALIAS])
+        plan = executor.migration_plan(executor.loader.graph.leaf_nodes())
+        if plan:
+            apps_waiting_migration = sorted({migration.app_label for migration, backwards in plan})
+            messages.add_message(
+                request,
+                messages.ERROR,
+                mark_safe(
+                    _("\nYou have %(unapplied_migration_count)s unapplied migration(s). "
```
Comment (atodorov) [inline tcms/core/middleware.py:49]: Remove "\n" from the message.

Comment (atodorov) [reply tcms/core/middleware.py:49]: Cut the message at the first sentence and add a link to https://kiwitcms.readthedocs.io/en/latest/installing_docker.html#initial-configuration-of-running-container similar to the message about Base URL. We want to sent the user to the appropriate documentation.

File: tcms/settings/common.py [hunk 2]
```diff
@@ -130,6 +130,7 @@
     'global_login_required.GlobalLoginRequiredMiddleware',
     'simple_history.middleware.HistoryRequestMiddleware',
     'tcms.core.middleware.CheckSettingsMiddleware',
+    'tcms.core.middleware.CheckUnappliedMigrationsMiddleware'
```
Comment (atodorov) [inline tcms/settings/common.py:133]: Testing will need the migrations from Django applied b/c we're looking for Site & later User and if no migration are applied everything starts failing even before we get to the new middleware.

File: tcms/core/middleware.py [hunk 3]
```diff
@@ -32,3 +34,20 @@ def process_request(self, request):
...
                 )
             )
+
+
+class CheckUnappliedMigrationsMiddleware(MiddlewareMixin):
+    def process_request(self, request):
+        executor = MigrationExecutor(connections[DEFAULT_DB_ALIAS])
+        plan = executor.migration_plan(executor.loader.graph.leaf_nodes())
+        if plan:
+            apps_waiting_migration = sorted({migration.app_label for migration, backwards in plan})
```
Comment (atodorov) [inline tcms/core/middleware.py:44]: this also prints I have 57 unapplied migrrations when all of them are actually applied. See how `showmigrations` works.

Comment (brymut) [reply tcms/core/middleware.py:44]: Ahh, didn't notice that, will switch to `MigrationLoader` from `showmigrations`.

File: tcms/settings/common.py (hunk 2 above)
Comment (brymut) [reply tcms/settings/common.py:133]: I was confused about how you meant for @ skipUnless to be applied here so I tried out setting up a small test app that we could use test with and without unapplied migrations by overriding the settings. I'm thinking this could work?

Comment (atodorov) [reply tcms/settings/common.py:133]: It will work but it is not what we want to do here. The existing test suite is extensive as it is, no need to complicate things even more.  With GitHub Actions you have the ability for custom setup and custom job definitions, with @skipUnless you can skip test classes/test methods based on conditions.

So you can have a new GH job definition which applies only some migartions as a bare minimum so the app doesn't crash; possibly define some ENV variable to indicate you want to run the tests about this particular middleware and then tell manage.py test to execute only tests for this middleware (b/c the rest will fail and are not of interest here).  You don't need to introduce a test app or dummy migrations.

File: tcms/core/middleware.py [hunk 4]
```diff
@@ -29,6 +30,26 @@ def process_request(self, request):
...
+        doc_url = "https://kiwitcms.readthedocs.io/en/latest/\
+installing_docker.html#initial-configuration-of-running-container"
+        executor = MigrationExecutor(connections[DEFAULT_DB_ALIAS])
+        plan = executor.migration_plan(executor.loader.graph.leaf_nodes())
+        if plan:
+            messages.add_message(
+                request,
+                messages.ERROR,
+                mark_safe(
+                    _('You have %(unapplied_migration_count)s unapplied migration(s).'
```
Comment (atodorov) [inline tcms/core/middleware.py:49]: missing space after the dot.

File: tcms/settings/common.py (hunk 2 above)
Comment (brymut) [reply tcms/settings/common.py:133]: I think instead of introducing a new ENV var, Django's test runner lets you define the specific test you want to run via the command.

Comment (brymut) [reply tcms/settings/common.py:133]: I'm having issues with this. The test runner automatically applies migrations against the test database.

File: tcms/core/middleware.py [hunk 5]
```diff
@@ -18,7 +20,6 @@ class CheckSettingsMiddleware(MiddlewareMixin):
     def process_request(self, request):
         doc_url = 'https://kiwitcms.readthedocs.io/en/latest/admin.html#configure-kiwi-s-base-url'
         site = Site.objects.get(pk=settings.SITE_ID)
-
```
Comment (atodorov) [inline tcms/core/middleware.py:21]: unrelated change, revert it.

File: tcms/core/middleware.py [hunk 6]
```diff
@@ -29,6 +30,26 @@ def process_request(self, request):
                       '<a href="%(admin_url)s">change it</a>') % {
                           'doc_url': doc_url,
                           'admin_url': reverse('admin:sites_site_change', args=[site.pk])
-                      }
+                          }
+                )
+            )
```
Comment (atodorov) [inline tcms/core/middleware.py:35]: these also seem unrelated. Why do we need them ?

File: tcms/core/middleware.py [hunk 7]
```diff
@@ -29,6 +30,26 @@ def process_request(self, request):
...
+
+
+class CheckUnappliedMigrationsMiddleware(MiddlewareMixin):
+    def process_request(self, request):
+        doc_url = "https://kiwitcms.readthedocs.io/en/latest/\
+installing_docker.html#initial-configuration-of-running-container"
```
Comment (atodorov) [inline tcms/core/middleware.py:41]: why not use multi-line string concatenation like it was before ? Also used on lines 49-50 below.

File: tcms/core/tests/migrations/models.py [hunk 8]
```diff
@@ -0,0 +1,5 @@
+from django.db import models
```
Comment (atodorov) [inline tcms/core/tests/migrations/models.py:1]: all of these need to be gone.

File: tcms/settings/common.py (hunk 2 above)
Comment (atodorov) [reply tcms/settings/common.py:133]: > I think instead of introducing a new ENV var, Django's test runner lets you define the specific test you want to run via the command.

that's true, but when we run all of the other tests (e.g. with MariaDB, Postgres, SQLite) we don't specify anything to the test runner so it needs a way to know how to exclude/skip the special test for migrations middleware. The easiest way to do this with ENV variables. The same way it is done when running tests for the internal bug tracker functionality or for external bug trackers.

Comment (atodorov) [reply tcms/settings/common.py:133]: >  I'm having issues with this. The test runner automatically applies migrations against the test database.

Well show me how you are starting your test job then? I don't see a separate test job in this PR yet.

File: tcms/core/middleware.py (hunk 7 above)
Comment (brymut) [reply tcms/core/middleware.py:41]: I'm confused, I had changed based on this feedback. https://github.com/kiwitcms/Kiwi/pull/1752#discussion_r444497884

File: tcms/core/middleware.py (hunk 6 above)
Comment (brymut) [reply tcms/core/middleware.py:35]: I had noticed that pylint complained about `Wrong hanging indentation.pylint(bad-continuation)`
I had been trying to ignore the error because my editor formats the code on saving (pylint). I unintentionally added it on the last push, will remove.

File: tcms/settings/common.py (hunk 2 above)
Comment (brymut) [reply tcms/settings/common.py:133]: Sorry, hadn't pushed yet because I hadn't got to a solution yet. Will push

File: .github/workflows/testing.yml [hunk 9]
```diff
@@ -35,6 +35,37 @@ jobs:
...
+    - name: Install Python dependencies
+      run: |
+        sudo apt-get install libkrb5-dev gettext
+        pip install -r requirements/devel.txt
+        pushd tcms/ && npm install && popd
+
+    - name: Run test
+      run: |
+        export LANG=en-us
+        ./manage.py migrate --settings=$(DJANGO_SETTINGS_MODULE) auth
```
Comment (atodorov) [inline .github/workflows/testing.yml:62]: This variable isn't defined but you can hard-code to tcms.settings.test

File: .github/workflows/testing.yml [hunk 10]
```diff
@@ -35,6 +35,37 @@ jobs:
...
+        ./manage.py migrate --settings=$(DJANGO_SETTINGS_MODULE) sites
+        ./manage.py migrate --settings=$(DJANGO_SETTINGS_MODULE) sessions
+
+        export TEST_CHECK_UNAPPLIED_MIGRATIONS_MIDDLEWARE=1
+        make test
```
Comment (atodorov) [inline .github/workflows/testing.yml:67]: This wants to execute all possible tests which will fail because 90% of the DB tables are missing.  See how we run the tests in `integration_bugtracker.yml` for hints how to tell the test runner to only execute the middleware test.

File: tcms/core/middleware.py [hunk 11]
```diff
@@ -32,3 +34,23 @@ def process_request(self, request):
                       }
                 )
             )
+
+
+class CheckUnappliedMigrationsMiddleware(MiddlewareMixin):
+    def process_request(self, request):
+        doc_url = ('https://kiwitcms.readthedocs.io/en/latest/installing_'
+                   'docker.html#initial-configuration-of-running-container')
```
Comment (atodorov) [inline tcms/core/middleware.py:42]: If you use the following syntax:

```
"line 1"
"line 2"
```

both strings will be concatenated together and there is no need to use brackets. Maybe that works with single quotes as well, IDK - I've never tried it.

Also try splitting the URL along its natural dividers like `/`

File: tcms/core/tests/test_middleware.py [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+from django.contrib.messages import get_messages
+
+
+@unittest.skipUnless(
+    os.getenv('TEST_CHECK_UNAPPLIED_MIGRATIONS_MIDDLEWARE'),
+    'CheckUnappliedMigrationsMiddleware testing not enabled')
+class TestCheckUnappliedMigrationsMiddleware(TransactionTestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.unapplied_migration_message = ('You have 41 unapplied migration(s)'
```
Comment (atodorov) [inline tcms/core/tests/test_middleware.py:13]: the number will change both when we add our own migrations and when Django or any of the other dependencies does, which will break this test.  Semi-auto upgrades like the ones with PyUP bot and Dependabot will become more fragile and require manual adjusting of this value.

Instead try asserting on parts of the message + the URL, it should be enough.

File: tcms/core/tests/test_middleware.py [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+                                           '. See <a href="https://kiwitcms.'
+                                           'readthedocs.io/en/latest/installin'
+                                           'g_docker.html#initial-configuratio'
+                                           'n-of-running-container">documentat'
+                                           'ion</a>')
```
Comment (atodorov) [inline tcms/core/tests/test_middleware.py:18]: Same comment for brackets and URL splitting as above.

File: tcms/core/tests/test_middleware.py [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+
+    def test_unapplied_migrations(self):
+        response = self.client.get('/')
+        messages = list(get_messages(response.wsgi_request))
+        self.assertEqual(str(messages[0]), self.unapplied_migration_message)
```
Comment (atodorov) [inline tcms/core/tests/test_middleware.py:23]: response is either a response containing an HTML page or a redirect (see get(follow=True)). Which means you can use assertContains against the response to search for strings in the HTML text.

File: tcms/settings/common.py (hunk 2 above)
Comment (atodorov) [inline tcms/settings/common.py:133]: add trailing comma to minimize diff next time this list is updates.

File: tcms/core/tests/test_middleware.py (hunk 14 above)
Comment (brymut) [reply tcms/core/tests/test_middleware.py:23]: Why not just use the purpose built `get_messages`?

Comment (atodorov) [reply tcms/core/tests/test_middleware.py:23]: because we want to know the message is actually shown on the page the user will see and because this is an extra step which isn't needed in this case.

File: tcms/core/middleware.py (hunk 11 above)
Comment (brymut) [reply tcms/core/middleware.py:42]: do you mean like:
`doc_url = "https://kiwitcms.readthedocs.io/en/latest/"`
`"installing_docker.html#initial-configuration-of-running-container"`

As I'm getting a pylint error `String statement has no effect pylint(pointless-string-statement)` Is there an example I can refer to?

Comment (brymut) [reply tcms/core/middleware.py]: ohh, never mind, I missed the \

File: .github/workflows/testing.yml [hunk 15]
```diff
@@ -35,6 +35,37 @@ jobs:
...
+    - name: Run test
+      run: |
+        export LANG=en-us
+        ./manage.py migrate --settings=tcms.settings.test auth
+        ./manage.py migrate --settings=tcms.settings.test sites
+        ./manage.py migrate --settings=tcms.settings.test sessions
+
+        export TEST_CHECK_UNAPPLIED_MIGRATIONS_MIDDLEWARE=1
+        ./manage.py test -v2 --noinput --settings=tcms.settings.test tcms.core.tests.test_middleware.TestCheckUnappliedMigrationsMiddleware
+
```
Comment (atodorov) [inline .github/workflows/testing.yml:68]: Execute this with coverage and after it is done send the results to codecov.io

File: .github/workflows/testing.yml [hunk 16]
```diff
@@ -35,6 +35,37 @@ jobs:
           coverage report -m
           bash <(curl -s https://codecov.io/bash)
 
+  test_check_unapplied_migrations_middleware:
+    name: test CheckUnappliedMigrationsMiddleware
```
Comment (atodorov) [inline .github/workflows/testing.yml:39]: change name only to "middleware for unapplied migrations"

File: tcms/core/tests/test_middleware.py [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+@unittest.skipUnless(
+    os.getenv('TEST_CHECK_UNAPPLIED_MIGRATIONS_MIDDLEWARE'),
+    'CheckUnappliedMigrationsMiddleware testing not enabled')
+class TestCheckUnappliedMigrationsMiddleware(TransactionTestCase):
+    @classmethod
+    def setUpClass(cls):
+        cls.unapplied_migration_message = "unapplied migration(s). See "\
+            "<a href=\"https://kiwitcms.readthedocs.io/en/latest/"\
+            "installing_docker.html#initial-configuration-of-running-"\
+            "container\">documentation</a>"
```
Comment (atodorov) [inline tcms/core/tests/test_middleware.py:15]: Now that there is only 1 scenario you can move the expected message directly in the test method. We don't need extra setUp methods

File: tcms/core/middleware.py [hunk 18]
```diff
@@ -32,3 +34,24 @@ def process_request(self, request):
                       }
                 )
             )
+
+
+class CheckUnappliedMigrationsMiddleware(MiddlewareMixin):
+    def process_request(self, request):
+        doc_url = "https://kiwitcms.readthedocs.io/en/latest/"\
+                  "installing_docker.html"\
+                  "#initial-configuration-of-running-container"
```
Comment (atodorov) [inline tcms/core/middleware.py:43]: I don't like this syntax, try:

```
        doc_url = """https://kiwitcms.readthedocs.io/en/latest/
installing_docker.html#initial-configuration-of-running-container"""
```

looks better IMO and tools are happy.

**Comment by brymut** [comment]
What do we do next? As I'd mentioned before the test runner applies the migrations automatically, I think we need to introduce a new test migration in the setup class of the testcase.

File: tcms/core/tests/test_middleware.py [hunk 19]
```diff
@@ -0,0 +1,19 @@
...
+
+    def test_unapplied_migrations(self):
+        response = self.client.get('/', follow=True)
+        self.assertContains(response, self.unapplied_migration_message)
```
Comment (atodorov) [inline tcms/core/tests/test_middleware.py:19]: The new test is failing in CI, make sure that is passes. You can print the response conents here to figure out what page it is. 

I'm not sure how TransactionTestCase will work in this situation, almost everywhere else we use test.TestCase which inherits from TransactionTestCase. You may need to change as well, IDK.

Comment (brymut) [reply tcms/core/tests/test_middleware.py]: Not seeing the message in the response, only seeing the Base URL not configured message

**Comment by atodorov** [comment]
> What do we do next? As I'd mentioned before the test runner applies the migrations automatically, I think we need to introduce a new test migration in the setup class of the testcase.

Right, I didn't realize this before.  All of the `manage.py migrate` commands in the test job are useless b/c they configure the production DB, not the test one, which is separate.  The test DB is configured automatically from the test runner and before any tests are started. It is not very easy to override this b/c the code goes into several layers deep.


However it should be possible to `import *` from the test settings file and create a special settings file to be used only for the migrations tests - try including only the same 3 apps like before. Then in the test itself try using `modify_settings/override_settings` telling Django to append or use all of the available apps:
https://docs.djangoproject.com/en/3.0/topics/testing/tools/#django.test.SimpleTestCase.modify_settings

In theory this should create a DB with a minimal number of migrations applied and then tell Django that there are more installed apps which will be missing their migrations and the test should pass.

**Comment by brymut** [comment]
> `modify_settings/override_settings` telling Django to append or use all of the available apps:
I'm having trouble with this. Could you please have a look.

**Comment by brymut** [comment]
Also, could you consider adding a test migration app instead like in #1767 .I feel like its less complicated to follow, more predictable and easier to extend in case we would like to add more migration-related testing in the future.

**Comment by atodorov** [comment]
test_middleware.py

```diff
+from django.core.management import call_command
 
 
 @unittest.skipUnless(
     os.getenv('TEST_CHECK_UNAPPLIED_MIGRATIONS_MIDDLEWARE'),
     'CheckUnappliedMigrationsMiddleware testing not enabled')
-class TestCheckUnappliedMigrationsMiddleware(test.TestCase):
-    @test.override_settings(INSTALLED_APPS=INSTALLED_APPS)
+class TestCheckUnappliedMigrationsMiddleware(test.TransactionTestCase):
     def test_unapplied_migrations(self):
+        call_command(
+            'migrate',
+            'bugs',
+            'zero',
+            verbosity=2,
+            interactive=False,
+        )
+
+
```

together with

```diff
diff --git a/tcms/bugs/migrations/0002_add_permissions.py b/tcms/bugs/migrations/0002_add_permissions.py
index 049ddf6..e0c2cf3 100644
--- a/tcms/bugs/migrations/0002_add_permissions.py
+++ b/tcms/bugs/migrations/0002_add_permissions.py
@@ -15,6 +15,15 @@ def forwards_add_perms(apps, schema_editor):
     tester.permissions.add(*app_perms)
 
 
+def backwards(apps, schema_editor):
+    Group = apps.get_model('auth', 'Group')
+    Permission = apps.get_model('auth', 'Permission')
+
+    tester = Group.objects.get(name='Tester')
+    app_perms = Permission.objects.filter(content_type__app_label__contains='bugs')
+    tester.permissions.remove(*app_perms)
+
+
 class Migration(migrations.Migration):
     dependencies = [
         ('bugs', '0001_initial'),
@@ -22,5 +31,5 @@ class Migration(migrations.Migration):
     ]
 
     operations = [
-        migrations.RunPython(forwards_add_perms),
+        migrations.RunPython(forwards_add_perms, backwards),
     ]
```

Will get this going. It is really as simple as calling migrate from within the test and reverting some of the migrations. The trouble is our migrations are not well written in a way which could be reverted hence the additional backwards migration to remove the permissions added in the forwards step. I am opening a new issue to create a linter for these migrations so we can fix them.

The rest is cleanup and formatting of those long strings which don't want to play nicely with the tools. After all I think your original implementation with the quotes and backslash was the most appropriate one but I didn't pay much attention to it. What I wanted to do I think is only possible if the string is passed as an argument to a function. E.g. this works for me and seems relatively nice

```diff
 class CheckUnappliedMigrationsMiddleware(MiddlewareMixin):
     def process_request(self, request):
-        doc_url = """https://kiwitcms.readthedocs.io/en/latest/
-installing_docker.html#initial-configuration-of-running-container"""
+        doc_url = 'https://kiwitcms.readthedocs.io/en/latest/' \
+                  'installing_docker.html#initial-configuration-of-running-container'
```

**Comment by brymut** [comment]
Doesn't seem to work, getting `django.db.migrations.exceptions.IrreversibleError: Operation <RunPython <function forwards_add_perms at 0x7f4102425950>> in bugs.0002_add_permissions is not reversible`
